# Laboratorio 2
## Miembros del equipo
- Ricardo Méndez 21289
- Sara Echeverría 21371
- Melissa Pérez 21385
- Francisco Castillo 21562

# Task 1

#### ¿Por qué el modelo de Naive Bayes se le considera “naive”?

La teoría de Bayes asume que los eventos son dependientes entre sí. Al implementarlo con un acercamiento _naive_ implica que asumimos que los eventos son independientes entre sí. Donde, básicamente, describe que una característica de un conjunto de datos no está relacionada con otra de ellas. En otras palabras, es por su simplicidad de la independencia condicional entre las características o atributos que describen los datos. [(Shoba, 2018)](https://www.sciencedirect.com/science/article/abs/pii/S0169716118300191)

#### Explique la formulación matemática que se busca optimizar en Support Vector Machine, además responda ¿cómo funciona el truco del Kernel para este modelo? 
Las máquinas de soporte vectorial buscan encontrar una frontera que maximice el margen de separation para la classification de los datos. La prediction de los datos se realiza con la siguiente formula:
$$
y_i = sign(w^Tx_i+b)
$$
El margen se optimiza a traves de la distancia perpendicular desde la frontera hacia al punto mas cercano de cada clase; es decir, maximizamos esta distancia. Esto se resuelve con multiplicadores de Lagrange (Tantos multiplicadores como cantidad de puntos tenemos)
$$
argmin_{w,ɑ} \frac{1}{2}w^Tw- \sum_{n,m=1}^{N}ɑ_mɑ_ny_my_nx_n,x_m
$$
Tras resolver el problema de optimización, se obtiene la siguiente formula:
$$
argmax_w \sum_{n=1}^{N}ɑ_n - \frac{1}{2} \sum_{n,m=1}^{N}ɑ_nɑ_my_my_n{x_n}^Tx_m
$$
Se puede apreciar que ya no dependemos de los pesos ($w$) y el bias ($b$), sino que dependemos de los multiplicadores de Lagrange ($ɑ$). Estos multiplicadores son los que nos permiten encontrar la frontera de decision.

Las predicciones ahora se hacen de la siguiente manera:
$$
y_i = sign(\sum_{n=1}^{N}ɑ_ny_nx_n^Tx_i+b)
$$
Los datos de $x_i$ solamente aparecen como un producto punto, el cual podemos representar con una función de kernel.
$$
y_i = sign(\sum_{n=1}^{N}ɑ_ny_nK(x_n,x_i)+b)
$$

EL utilizar esta function nos permite ingress el _truco del kernel_. Este truco nos permite mapear los datos a un espacio de mayor dimension, donde es más facial encontrar una frontera de decision que no es necesariamente lineal.  [(Ranjan, 2019)](https://towardsdatascience.com/truly-understanding-the-kernel-trick-1aeb11560769)

### ¿Qué tipo de ensemble learning es este modelo (Random Forest)?
Random Forest es un ensemble learning de tipo _Bootsrap Aggregation_ o _bagging_. Se utiliza para reducir la varianza de un modelo de machine learning a base de crear multiples modelos y combinarlos para obtener un modelo final. En un _Random Forest_ todos los modelos son árboles de decision. Cada arbol se entrena con diferentes subconjuntos de datos los cuales son combinados para obtener un modelo final. [(Sruthi, 2024)](https://www.analyticsvidhya.com/blog/2021/06/understanding-random-forest/)

### ¿Cuál es la idea general detrás de Random Forest?
Como se ha mencionado anteriormente, el objetivo es crear un modelo de aprendizaje que combine la salida de multiples arboles de decision para llegar a un resultado unico. 

### ¿Por qué se busca baja correlación entre los árboles de Random Forest?
Se busca que los _features_ tengan baja correlación para reducir la varianza y mejorar la precision de las predicciones. Si la correlación es alta entre estos arboles, es muy probable que cometan los mismos errores en la predicción, lo que no ofrecería ninguna mejora. [(Breiman & Cutler, s.f.)](https://www.stat.berkeley.edu/~breiman/RandomForests/cc_home.htm)

In [26]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

scaler = StandardScaler()

# Task 2 Naive Bayes: Clasificador de Mensajes Ham/Spam

Los datos de prueba han sido obtenidos de [este enlace](https://github.com/Sk70249/NLP-Spam-Ham-Classifier/blob/master/data/SMSSpamCollection.tsv)

# Task 2.1

# Task 2.2

# Task 2.3

# Task 2.4

# Task 3 Clasificación de Partidas de League of Legends

## Análisis Exploratorio

In [27]:
df = pd.read_csv('data/high_diamond_ranked_10min.csv')

In [28]:
df.head()

gameId  blueWins  blueWardsPlaced  blueWardsDestroyed  blueFirstBlood  \
0  4519157822         0               28                   2               1   
1  4523371949         0               12                   1               0   
2  4521474530         0               15                   0               0   
3  4524384067         0               43                   1               0   
4  4436033771         0               75                   4               0   

   blueKills  blueDeaths  blueAssists  blueEliteMonsters  blueDragons  ...  \
0          9           6           11                  0            0  ...   
1          5           5            5                  0            0  ...   
2          7          11            4                  1            1  ...   
3          4           5            5                  1            0  ...   
4          6           6            6                  0            0  ...   

   redTowersDestroyed  redTotalGold  redAvgLevel  redTotalExperience  \
0                   0         16567          6.8               17047   
1                   1         17620          6.8               17438   
2                   0         17285          6.8               17254   
3                   0         16478          7.0               17961   
4                   0         17404          7.0               18313   

   redTotalMinionsKilled  redTotalJungleMinionsKilled  redGoldDiff  \
0                    197                           55         -643   
1                    240                           52         2908   
2                    203                           28         1172   
3                    235                           47         1321   
4                    225                           67         1004   

   redExperienceDiff  redCSPerMin  redGoldPerMin  
0                  8         19.7         1656.7  
1               1173         24.0         1762.0  
2               1033         20.3         1728.5  
3                  7         23.5         1647.8  
4               -230         22.5         1740.4  

[5 rows x 40 columns]

In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9879 entries, 0 to 9878
Data columns (total 40 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   gameId                        9879 non-null   int64  
 1   blueWins                      9879 non-null   int64  
 2   blueWardsPlaced               9879 non-null   int64  
 3   blueWardsDestroyed            9879 non-null   int64  
 4   blueFirstBlood                9879 non-null   int64  
 5   blueKills                     9879 non-null   int64  
 6   blueDeaths                    9879 non-null   int64  
 7   blueAssists                   9879 non-null   int64  
 8   blueEliteMonsters             9879 non-null   int64  
 9   blueDragons                   9879 non-null   int64  
 10  blueHeralds                   9879 non-null   int64  
 11  blueTowersDestroyed           9879 non-null   int64  
 12  blueTotalGold                 9879 non-null   int64  
 13  blu

In [30]:
df.describe()

gameId     blueWins  blueWardsPlaced  blueWardsDestroyed  \
count  9.879000e+03  9879.000000      9879.000000         9879.000000   
mean   4.500084e+09     0.499038        22.288288            2.824881   
std    2.757328e+07     0.500024        18.019177            2.174998   
min    4.295358e+09     0.000000         5.000000            0.000000   
25%    4.483301e+09     0.000000        14.000000            1.000000   
50%    4.510920e+09     0.000000        16.000000            3.000000   
75%    4.521733e+09     1.000000        20.000000            4.000000   
max    4.527991e+09     1.000000       250.000000           27.000000   

       blueFirstBlood    blueKills   blueDeaths  blueAssists  \
count     9879.000000  9879.000000  9879.000000  9879.000000   
mean         0.504808     6.183925     6.137666     6.645106   
std          0.500002     3.011028     2.933818     4.064520   
min          0.000000     0.000000     0.000000     0.000000   
25%          0.000000     4.000000     4.000000     4.000000   
50%          1.000000     6.000000     6.000000     6.000000   
75%          1.000000     8.000000     8.000000     9.000000   
max          1.000000    22.000000    22.000000    29.000000   

       blueEliteMonsters  blueDragons  ...  redTowersDestroyed  redTotalGold  \
count        9879.000000  9879.000000  ...         9879.000000   9879.000000   
mean            0.549954     0.361980  ...            0.043021  16489.041401   
std             0.625527     0.480597  ...            0.216900   1490.888406   
min             0.000000     0.000000  ...            0.000000  11212.000000   
25%             0.000000     0.000000  ...            0.000000  15427.500000   
50%             0.000000     0.000000  ...            0.000000  16378.000000   
75%             1.000000     1.000000  ...            0.000000  17418.500000   
max             2.000000     1.000000  ...            2.000000  22732.000000   

       redAvgLevel  redTotalExperience  redTotalMinionsKilled  \
count  9879.000000         9879.000000            9879.000000   
mean      6.925316        17961.730438             217.349226   
std       0.305311         1198.583912              21.911668   
min       4.800000        10465.000000             107.000000   
25%       6.800000        17209.500000             203.000000   
50%       7.000000        17974.000000             218.000000   
75%       7.200000        18764.500000             233.000000   
max       8.200000        22269.000000             289.000000   

       redTotalJungleMinionsKilled   redGoldDiff  redExperienceDiff  \
count                  9879.000000   9879.000000        9879.000000   
mean                     51.313088    -14.414111          33.620306   
std                      10.027885   2453.349179        1920.370438   
min                       4.000000 -11467.000000       -8348.000000   
25%                      44.000000  -1596.000000       -1212.000000   
50%                      51.000000    -14.000000          28.000000   
75%                      57.000000   1585.500000        1290.500000   
max                      92.000000  10830.000000        9333.000000   

       redCSPerMin  redGoldPerMin  
count  9879.000000    9879.000000  
mean     21.734923    1648.904140  
std       2.191167     149.088841  
min      10.700000    1121.200000  
25%      20.300000    1542.750000  
50%      21.800000    1637.800000  
75%      23.300000    1741.850000  
max      28.900000    2273.200000  

[8 rows x 40 columns]

In [31]:
df['blueWins'].value_counts(normalize=True)

blueWins
0    0.500962
1    0.499038
Name: proportion, dtype: float64

In [32]:
features = df.drop(columns=['blueWins', 'gameId'])
target = df['blueWins']

In [33]:
# Scale everything
features = pd.DataFrame(scaler.fit_transform(features), columns=features.columns)

In [34]:
features.describe()

blueWardsPlaced  blueWardsDestroyed  blueFirstBlood     blueKills  \
count     9.879000e+03        9.879000e+03    9.879000e+03  9.879000e+03   
mean     -2.876982e-17        5.034719e-18   -1.078868e-18  1.125619e-16   
std       1.000051e+00        1.000051e+00    1.000051e+00  1.000051e+00   
min      -9.594869e-01       -1.298863e+00   -1.009663e+00 -2.053863e+00   
25%      -4.599937e-01       -8.390689e-01   -1.009663e+00 -7.253456e-01   
50%      -3.489952e-01        8.051859e-02    9.904294e-01 -6.108705e-02   
75%      -1.269983e-01        5.403123e-01    9.904294e-01  6.031716e-01   
max       1.263783e+01        1.111557e+01    9.904294e-01  5.252982e+00   

         blueDeaths   blueAssists  blueEliteMonsters   blueDragons  \
count  9.879000e+03  9.879000e+03       9.879000e+03  9.879000e+03   
mean  -1.179563e-16 -1.111234e-16       3.308530e-17 -8.702872e-17   
std    1.000051e+00  1.000051e+00       1.000051e+00  1.000051e+00   
min   -2.092146e+00 -1.634988e+00      -8.792310e-01 -7.532257e-01   
25%   -7.286663e-01 -6.508123e-01      -8.792310e-01 -7.532257e-01   
50%   -4.692613e-02 -1.587244e-01      -8.792310e-01 -7.532257e-01   
75%    6.348140e-01  5.794075e-01       7.195032e-01  1.327623e+00   
max    5.406995e+00  5.500287e+00       2.318237e+00  1.327623e+00   

        blueHeralds  blueTowersDestroyed  ...  redTowersDestroyed  \
count  9.879000e+03         9.879000e+03  ...        9.879000e+03   
mean  -5.753965e-18        -2.733133e-17  ...        2.589284e-17   
std    1.000051e+00         1.000051e+00  ...        1.000051e+00   
min   -4.811324e-01        -2.104390e-01  ...       -1.983529e-01   
25%   -4.811324e-01        -2.104390e-01  ...       -1.983529e-01   
50%   -4.811324e-01        -2.104390e-01  ...       -1.983529e-01   
75%   -4.811324e-01        -2.104390e-01  ...       -1.983529e-01   
max    2.078430e+00         1.615907e+01  ...        9.022956e+00   

       redTotalGold   redAvgLevel  redTotalExperience  redTotalMinionsKilled  \
count  9.879000e+03  9.879000e+03        9.879000e+03           9.879000e+03   
mean   1.146837e-15  1.444245e-15        3.394839e-16           5.897814e-16   
std    1.000051e+00  1.000051e+00        1.000051e+00           1.000051e+00   
min   -3.539707e+00 -6.961495e+00       -6.254973e+00          -5.036349e+00   
25%   -7.120554e-01 -4.104749e-01       -6.276311e-01          -6.549000e-01   
50%   -7.448379e-02  2.446271e-01        1.023723e-02           2.970140e-02   
75%    6.234576e-01  8.997291e-01        6.697989e-01           7.143028e-01   
max    4.187620e+00  4.175239e+00        3.593814e+00           3.270148e+00   

       redTotalJungleMinionsKilled   redGoldDiff  redExperienceDiff  \
count                 9.879000e+03  9.879000e+03       9.879000e+03   
mean                  8.055551e-17 -1.258680e-17       1.438491e-18   
std                   1.000051e+00  1.000051e+00       1.000051e+00   
min                  -4.718391e+00 -4.668380e+00      -4.364806e+00   
25%                  -7.293122e-01 -6.446966e-01      -6.486683e-01   
50%                  -3.122336e-02  1.688026e-04      -2.926826e-03   
75%                   5.671385e-01  6.521677e-01       6.545317e-01   
max                   4.057583e+00  4.420473e+00       4.842738e+00   

        redCSPerMin  redGoldPerMin  
count  9.879000e+03   9.879000e+03  
mean  -1.352901e-15  -6.883180e-16  
std    1.000051e+00   1.000051e+00  
min   -5.036349e+00  -3.539707e+00  
25%   -6.549000e-01  -7.120554e-01  
50%    2.970140e-02  -7.448379e-02  
75%    7.143028e-01   6.234576e-01  
max    3.270148e+00   4.187620e+00  

[8 rows x 38 columns]

### Pasos del análisis exploratorio

#### Encoding
- No se ha realizado encoding, ya que no hay variables categóricas

#### Balanceo
- No se ha realizado balanceo, ya que el dataset está balanceado

#### Escalamiento
- Se ha realizado escalamiento de los datos, ya que las variables tienen diferentes escalas

#### Selección de Variables
- No se ha realizado selección de variables, ya que todas las variables son relevantes

In [35]:
X_train, X_test, y_train, y_test = train_test_split(features, target, test_size=0.2, random_state=7)

## Task 3.1 Support Vector Machines

## Task 3.2 Árboles de Decisión

## Task 3.3 Comparación